context-propagation-java8/src/main/java/nl/talsmasoftware/context/functions/BiConsumerWithContext.java

```diff
@@ -33,20 +33,24 @@
  *
  * @author Sjoerd Talsma
  */
-public class BiConsumerWithContext<T, U> extends WrapperWithContext<BiConsumer<T, U>> implements BiConsumer<T, U> {
+public class BiConsumerWithContext<T, U> extends Java8WrapperWithContext<BiConsumer<T, U>> implements BiConsumer<T, U> {
     private static final Logger LOGGER = Logger.getLogger(BiConsumerWithContext.class.getName());
 
     public BiConsumerWithContext(ContextSnapshot snapshot, BiConsumer<T, U> delegate) {
         this(snapshot, delegate, null);
     }
 
     public BiConsumerWithContext(ContextSnapshot snapshot, BiConsumer<T, U> delegate, Consumer<ContextSnapshot> consumer) {
-        super(snapshot, delegate, consumer == null ? null : consumer::accept);
+        super(snapshot, delegate, consumer);
+    }
+
+    protected BiConsumerWithContext(Supplier<ContextSnapshot> supplier, BiConsumer<T, U> delegate, Consumer<ContextSnapshot> consumer) {
+        super(supplier, delegate, consumer);
     }
 
     @Override
     public void accept(T t, U u) {
-        try (Context<Void> context = snapshot.reactivate()) {
+        try (Context<Void> context = snapshot().reactivate()) {
             try {
```

Comment:
Is it possible to use only one 'try'?

Reply:
Unfortunately not.

We have a requirement that a new snapshot is captured and returned to the caller _after_ the function is called, but _before_ the reactivated context is closed (otherwise there's a good chance that we capture and return an empty snapshot).

The [Oracle documentation](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) explains the ordering of _try-with-resources_ and _finally_ statements:
> Note: A try-with-resources statement can have catch and finally blocks just like an ordinary try statement. In a try-with-resources statement, any catch or finally block is run after the resources declared have been closed.

Therefore, we require both an inner-try for _capturing_ and an outer try for _closing_ (unfortunately).